### PR TITLE
feat[fork]: add set_code, get_storage, set_storage to boa.env

### LINKED ISSUE #204: Provide a higher-level API to manipulate forked-network state
I would like a helper method to override account state, similar to `hardhat_setStorageAt` feature by hardhat. The use-case is when running in forking mode, I would like my tests to run thru various scenarios which is only possible by manipulating state of other contracts.

For now, I am able to do this by going lower-level and doing these against pyEVM directly. Here is an example where I set the balance of a given account to 5000 stETH

```python
    abi_encoded = eth_abi.encode(['address', 'uint256'], [trader, 0])
    storage_slot = Web3.solidity_keccak(["bytes"], ["0x" + abi_encoded.hex()])

    boa.env.evm.vm.state.set_storage(
        boa.util.abi.Address(STETH).canonical_address,
        Web3.to_int(storage_slot),
        5000 * 10**18
    )
```

### PR DESCRIPTION
**Pull request by DanielSchiavini**
Fixes #204

image

### CODE REVIEW

File: boa/environment.py [hunk 1]
```diff
@@ -311,9 +311,22 @@ def _hook_trace_computation(self, computation, contract=None):
...
-    def get_code(self, address):
+    def get_code(self, address: _AddressType) -> bytes:
         return self.evm.get_code(Address(address))
 
+    def set_code(self, address: _AddressType, code: bytes) -> None:
+        assert self.evm.is_forked, "The EVM is not forked, cannot set code"
+        self.evm.set_code(Address(address), code)
+
+    def get_storage(
+        self, address: _AddressType, slot: int, from_journal: bool = True
```
Comment (charles-cooper) [inline boa/environment.py:322]: i don't think we need `from_journal`, as it exposes py-evm internals most users are not concerned with

File: boa/environment.py [hunk 2]
```diff
@@ -311,9 +311,22 @@ def _hook_trace_computation(self, computation, contract=None):
             child_contract = self._lookup_contract_fast(child.msg.code_address)
             self._hook_trace_computation(child, child_contract)
 
-    def get_code(self, address):
+    def get_code(self, address: _AddressType) -> bytes:
         return self.evm.get_code(Address(address))
 
+    def set_code(self, address: _AddressType, code: bytes) -> None:
+        assert self.evm.is_forked, "The EVM is not forked, cannot set code"
```
Comment (charles-cooper) [inline boa/environment.py:318]: why do we care if the evm is forked? imo it's even more useful in non-fork mode

File: boa/environment.py [hunk 3]
```diff
@@ -311,9 +311,22 @@ def _hook_trace_computation(self, computation, contract=None):
...
+        self.evm.set_code(Address(address), code)
+
+    def get_storage(
+        self, address: _AddressType, slot: int, from_journal: bool = True
+    ) -> int:
+        return self.evm.get_storage(Address(address), slot, from_journal)
+
+    def set_storage(self, address: _AddressType, slot: int, value: int) -> None:
+        assert self.evm.is_forked, "The EVM is not forked, cannot set storage"
```
Comment (charles-cooper) [inline boa/environment.py:327]: ditto https://github.com/vyperlang/titanoboa/pull/210/files#r1593856000

File: tests/unitary/test_fork_utils.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+def test_code():
+    boa.env.evm._set_account_db_class(AccountDBFork)
+    assert boa.env.get_code(address) == b""
+    code = b"some test code"
+    boa.env.set_code(address, code)
+    assert boa.env.get_code(address) == code
+
+
+def test_storage(monkeypatch):
```
Comment (charles-cooper) [inline tests/unitary/test_fork_utils.py:27]: doesn't use monkeypatch

File: boa/vm/fork.py [hunk 5]
```diff
@@ -120,7 +120,10 @@ def fetch_multi(self, payload):
             # fetch_multi is called only with the missing payloads
             # map the results back to the original indices
             for result_ix, rpc_result in enumerate(self._rpc.fetch_multi(batch)):
-                key, item_ix = keys[result_ix]
+                try:
+                    key, item_ix = keys[result_ix]
+                except IndexError:
```
Comment (charles-cooper) [inline boa/vm/fork.py:125]: looks spooky

Comment (DanielSchiavini) [reply boa/vm/fork.py:125]: sorry this shouldn't have been committed

**Review by charles-cooper — APPROVED** [review]

**Pull request merged**